File: ripperdoc/wetware/directive/scripts/show_mode.py
#!/usr/bin/env python3
"""Show a directive body — print skills/<value>/directive.md verbatim.

Usage: show_mode.py <value>

The per-mode echo skills (/swarm, /ask, /auto) invoke this to display their own
directive without writing anything. It is read-only: it never touches
preemdeck.json. <value> must be a bare name — anything carrying a path separator
or a dot-segment is rejected, so it can't escape the skills dir (the same guard
inject_mode's load_mode_text uses). The body is printed exactly as it ships, no
framing. Same input → same bytes.

Exit codes:
  0  directive printed
  2  usage error, unsafe value, or no matching skills/<value>/directive.md
"""

from __future__ import annotations

import sys
from pathlib import Path

SKILLS_DIR = Path(__file__).resolve().parents[1] / "skills"


def available_modes() -> list[str]:
    """Sorted mode names — skill folders that ship a `directive.md`."""
    if not SKILLS_DIR.is_dir():
        return []
    return sorted(d.name for d in SKILLS_DIR.iterdir() if d.is_dir() and (d / "directive.md").is_file())


def main(argv: list[str]) -> int:
    modes = available_modes()
    listing = ", ".join(modes) or "none"
    if len(argv) != 1 or not argv[0].strip():
        print(f"usage: show_mode.py <value>   (values: {listing})", file=sys.stderr)
        return 2
    value = argv[0].strip()
    if Path(value).name != value:  # path separator or dot-segment — refuse to escape
        print(f"unsafe value {value!r}; available: {listing}", file=sys.stderr)
        return 2
    body = SKILLS_DIR / value / "directive.md"
    if not body.is_file():
        print(f"unknown value {value!r}; available: {listing}", file=sys.stderr)
        return 2
    sys.stdout.write(body.read_text(encoding="utf-8"))
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: ripperdoc/wetware/directive/scripts/show-mode.ts
#!/usr/bin/env -S preemdeck-bun
/**
 * show-mode.ts — print a directive body verbatim.
 *
 * Read-only: never touches preemdeck.json. Prints skills/<value>/directive.md
 * exactly as it ships, no framing. <value> must be a bare name (the same
 * anti-traversal guard inject-mode uses). Same input -> same bytes.
 *
 * Exit codes: 0 printed; 2 usage error, unsafe value, or no matching directive.md.
 */

import { readdir, readFile, stat } from "node:fs/promises"
import { dirname, join } from "node:path"
import { exists } from "../../../../lib/fs.ts"
import { pyName } from "./pyname.ts"

const SKILLS_DIR = join(dirname(import.meta.dir), "skills")

/** Sorted mode names — skill folders that ship a `directive.md`. */
export const availableModes = async (skillsDir: string): Promise<string[]> => {
  if (!(await exists(skillsDir)) || !(await stat(skillsDir)).isDirectory()) return []
  const names: string[] = []
  const entries = await readdir(skillsDir)
  for (const entry of entries) {
    const dir = join(skillsDir, entry)
    if (
      (await stat(dir)).isDirectory() &&
      (await exists(join(dir, "directive.md"))) &&
      (await stat(join(dir, "directive.md"))).isFile()
    ) {
      names.push(entry)
    }
  }
  return names.sort()
}

/**
 * The CLI entry: validate <value> against the bare-name guard, then print its
 * directive body verbatim. Returns the process exit code (0 printed; 2 on usage
 * error, an unsafe value, or no matching directive.md) so callers stay testable.
 */
export const main = async (
  argv: string[],
  skillsDir: string = SKILLS_DIR,
  write: (s: string) => void = (s) => process.stdout.write(s),
): Promise<number> => {
  const modes = await availableModes(skillsDir)
  const listing = modes.join(", ") || "none"
  if (argv.length !== 1 || !argv[0] || argv[0].trim() === "") {
    process.stderr.write(`usage: show-mode <value>   (values: ${listing})\n`)
    return 2
  }
  const value = (argv[0] as string).trim()
  if (pyName(value) !== value) {
    process.stderr.write(`unsafe value ${pyRepr(value)}; available: ${listing}\n`)
    return 2
  }
  const body = join(skillsDir, value, "directive.md")
  if (!(await exists(body)) || !(await stat(body)).isFile()) {
    process.stderr.write(`unknown value ${pyRepr(value)}; available: ${listing}\n`)
    return 2
  }
  write(await readFile(body, "utf8"))
  return 0
}

/** Render a string the way Python's `{value!r}` does for the common cases. */
const pyRepr = (value: string): string => {
  if (!value.includes("'") || value.includes('"')) return `'${value}'`
  return `"${value}"`
}

if (import.meta.main) {
  process.exit(await main(Bun.argv.slice(2)))
}
